Directory structure doc (#27233)
openssl/openssl: doc/directory-structure.md

# OpenSSL Repository Structure

This document provides an overview of the OpenSSL repository's directory structure, explaining what the major directories and files are, their purpose, and how they relate to each other.

## Repository Tree Overview

Here's a high-level view of the repository structure:

```

> openssl/

Author: Added missing directories to the tree view:
.ctags.d/ - ctags configuration
.github/ - GitHub-specific files
Configurations/ - Platform build configs
dev/ - Development utilities
exporters/ - Build export utilities
os-dep/ - OS-dependent code

  * Usage: `./demos/quic/server/server <port> <certificate-file> <key-file>`

### doc/

Documentation organized in multiple formats:

* `doc/man1/` - Manual pages for command-line tools
* `doc/man3/` - Manual pages for API functions
* `doc/man5/` - Manual pages for file formats
* `doc/man7/` - Manual pages for overviews and concepts
* `doc/internal/` - Developer documentation

### include/

Header files that are installed along with OpenSSL:

* `include/openssl/` - Public API headers 
* `include/crypto/` - Headers for internal cryptographic implementations
* `include/internal/` - Headers for other internal components

### providers/

The provider architecture (new in OpenSSL 3.0) implementations:

* `providers/default/` - The default provider
* `providers/fips/` - The FIPS-validated module
* `providers/legacy/` - Legacy algorithms

### ssl/

Implementation of SSL, TLS, DTLS, and QUIC protocols:

* `ssl/record/` - SSL/TLS record layer
* `ssl/statem/` - SSL/TLS state machine
* `ssl/quic/` - QUIC protocol implementation (available since version 3.2.0)

### test/

Comprehensive test suite, organized by component and feature.

> ## Build System Files

Author: Added exporters/ and os-dep/ to the "Build System Files" section with proper descriptions

* `build.info` - Files in these directories define what gets built
* `Configurations/` - Platform-specific build configurations
* `util/` - Build scripts and utilities

## External Modules

OpenSSL can include several external modules as Git submodules, including:

* `oqs-provider/` - Post-quantum cryptography provider
* `krb5/` - Kerberos support
* `cloudflare-quiche/` - QUIC implementation from Cloudflare (used as a reference for OpenSSL's own QUIC implementation)

> ## Development and Contributor Files

Author: Added missing directories to the tree view:
.ctags.d/ - ctags configuration
.github/ - GitHub-specific files
Configurations/ - Platform build configs
dev/ - Development utilities
exporters/ - Build export utilities
os-dep/ - OS-dependent code

* `.github/` - GitHub-specific files (workflows, issue templates)
* `CONTRIBUTING.md` - Guidelines for contributing to OpenSSL
* `HACKING.md` - Information on modifying OpenSSL source

## Stability Information

The API provided in the public headers (`include/openssl/`) is considered stable within major versions, while internal headers may change more frequently. The provider interface is also considered stable.

Components currently experiencing significant development:
* The QUIC implementation is still evolving
* The provider architecture was introduced in 3.0 and continues to mature

Long-standing stable components:
* Core cryptographic algorithms in `crypto/`
* The SSL/TLS protocol implementation (excluding QUIC)

## File Organization Patterns

OpenSSL follows consistent naming patterns:
* `.c` files contain implementations
* `.h` files are headers
* `.pod` files are documentation sources
* `*_local.h` files are for internal use within a component
* `build.info` describes what to build in each directory

> * Test files typically include the name of what they're testing

Author: Changed "Test files" to "Test filenames" for better precision

# OpenSSL Repository Structure

> This document provides an overview of the OpenSSL repository's directory structure, explaining what the major directories and files are, their purpose, and how they relate to each other.

Reviewer: please split the line

├── os-dep/       # Operating system dependent code
├── providers/    # Provider implementations (FIPS, etc.)
├── ssl/          # SSL/TLS implementation
├── test/         # Test suite
├── tools/        # Various utility tools
├── util/         # Build-related utilities
└── VMS/          # OpenVMS specific files
```

## Root-Level Files

The repository root contains many important files:

* `README.md` - The main README with project overview and links to documentation
* `INSTALL.md` - Detailed installation instructions for all platforms
* `Configure` - The main build configuration script
* `config` - A wrapper around Configure for Unix platforms
* `LICENSE.txt` - The Apache 2.0 license that applies to OpenSSL 3.0+
* `CHANGES.md` - A comprehensive changelog
* `NEWS.md` - A curated list of major changes in each release
* `VERSION.dat` - Contains the current version information
* `configdata.pm.in` - Template used during build configuration

Various platform-specific notes are also at the root level:
* `NOTES-*.md` - Platform-specific information (UNIX, WINDOWS, etc.)
* `README-*.md` - Additional READMEs for specific components (ENGINES, FIPS, etc.)

## Major Directories

### apps/

> Contains the OpenSSL command-line application source code. This is what you're using when you run the `openssl` command.

Reviewer: Please split the line.

Notable components:
* The `openssl s_client` tool supports QUIC via the `-quic` flag (since version 3.2.0)
* The `openssl s_server` tool does NOT support QUIC directly

### crypto/

> The heart of OpenSSL containing implementations of cryptographic algorithms. Inside you'll find subdirectories for different cryptographic functions. This is the code that ends up in `libcrypto`.

Reviewer: Please split the line.

* `crypto/aes/` - AES encryption implementation
* `crypto/bn/` - Big Number arithmetic library
* `crypto/evp/` - High-level cryptographic API
* `crypto/rsa/`, `crypto/ec/`, etc. - Various algorithm implementations

### demos/

Example code demonstrating how to use OpenSSL APIs:

* `demos/quic/` - QUIC protocol examples, including a dedicated server implementation

>   * `demos/quic/server/` - Contains a standalone QUIC server application that should be used instead of `s_server` for QUIC functionality

Reviewer: Please split the long lines

  * Usage: `./demos/quic/server/server <port> <certificate-file> <key-file>`

### doc/

Documentation organized in multiple formats:

* `doc/man1/` - Manual pages for command-line tools
* `doc/man3/` - Manual pages for API functions
* `doc/man5/` - Manual pages for file formats
* `doc/man7/` - Manual pages for overviews and concepts
* `doc/internal/` - Developer documentation

### include/

Header files that are installed along with OpenSSL:

* `include/openssl/` - Public API headers 
* `include/crypto/` - Headers for internal cryptographic implementations
* `include/internal/` - Headers for other internal components

### providers/

The provider architecture (new in OpenSSL 3.0) implementations:

* `providers/default/` - The default provider
* `providers/fips/` - The FIPS-validated module
* `providers/legacy/` - Legacy algorithms

### ssl/

> Implementation of SSL, TLS, DTLS, and QUIC protocols. This is the code that ends up in `libssl`.

Reviewer: Please split the line.

* `ssl/record/` - SSL/TLS record layer
* `ssl/statem/` - SSL/TLS state machine
* `ssl/quic/` - QUIC protocol implementation (available since version 3.2.0)

### test/

Comprehensive test suite, organized by component and feature.

## Build System Files

* `Configurations/` - Platform-specific build configurations
* `exporters/` - Utilities for exporting OpenSSL build information to various build systems
* `os-dep/` - Operating system dependent code and platform-specific implementations
* `util/` - Build scripts and utilities

## External Modules

OpenSSL includes several external modules as Git submodules, mainly used for testing, including:

* `oqs-provider/` - Post-quantum cryptography provider
* `krb5/` - Kerberos support

> * `cloudflare-quiche/` - QUIC implementation from Cloudflare (used as a reference for OpenSSL's own QUIC implementation)

Reviewer: Please split the long line

## Development and Contributor Files

* `.ctags.d/` - Configuration for ctags, enabling better code navigation and indexing for developers
* `.github/` - GitHub-specific files (workflows, issue templates)
* `CONTRIBUTING.md` - Guidelines for contributing to OpenSSL
* `HACKING.md` - Information on adding custom modifications to OpenSSL source
* `dev/` - Development utilities, helper scripts, and tools used during OpenSSL development

## Stability Information

> The API provided in the public headers (`include/openssl/`) is considered stable within major versions, while internal headers may change more frequently. The provider interface is also considered stable.

Reviewer: Please split

Components currently experiencing significant development:
* The QUIC implementation is still evolving
* The provider architecture was introduced in 3.0 and continues to mature

Long-standing stable components:
* Core cryptographic algorithms in `crypto/`
* The SSL/TLS protocol implementation (excluding QUIC)

## File Organization Patterns

OpenSSL follows consistent naming patterns:
* `.c` files contain implementations
* `.h` files are headers
* `.pod` files are documentation sources
* `*_local.h` files are for internal use within a component
* `build.info` describes what to build in each directory
* Test filenames typically include the name of what they're testing

> This overview should help new contributors better understand the repository structure and navigate the codebase more effectively. 

Reviewer: And here

# OpenSSL Repository Structure

This document provides an overview of the OpenSSL repository's directory structure, explaining what the major directories and files are, their purpose, and how they relate to each other.

## Repository Tree Overview

Here's a high-level view of the repository structure:

```
openssl/
├── .ctags.d/     # ctags configuration for code navigation
├── .github/      # GitHub-specific files and workflows
├── apps/         # OpenSSL command line applications
├── Configurations/ # Build configuration for different platforms

> ├── crypto/       # Core cryptographic implementations

Author: @levitte Done!

├── demos/        # Example code demonstrating OpenSSL usage
├── dev/          # Development utilities and scripts
├── doc/          # Documentation
├── engines/      # Legacy engine implementations
├── exporters/    # Build system export utilities
├── external/     # External dependencies
├── fuzz/         # Fuzzing tools and configuration

> ├── include/      # Public header files

Reviewer 2: I see that you're going into more details further down, so just to avoid confusion...

```suggestion
├── include/      # Header files
```

├── ms/           # Microsoft Windows specific files
├── os-dep/       # Operating system dependent code
├── providers/    # Provider implementations (FIPS, etc.)

> ├── ssl/          # SSL/TLS implementation

Reviewer 2: You might want to say that this is the code that ends up in `libssl`

Author: @levitte Done!

├── test/         # Test suite
├── tools/        # Various utility tools
├── util/         # Build-related utilities
└── VMS/          # OpenVMS specific files
```

## Root-Level Files

The repository root contains many important files:

* `README.md` - The main README with project overview and links to documentation
* `INSTALL.md` - Detailed installation instructions for all platforms
* `Configure` - The main build configuration script
* `config` - A wrapper around Configure for Unix platforms
* `LICENSE.txt` - The Apache 2.0 license that applies to OpenSSL 3.0+
* `CHANGES.md` - A comprehensive changelog
* `NEWS.md` - A curated list of major changes in each release
* `VERSION.dat` - Contains the current version information

> * `*.pc.in` and `configdata.pm.in` - Templates used during build configuration

Reviewer 2: `*.pc.in` are in `exporters/`

```suggestion
* `configdata.pm.in` - Template used during build configuration
```

  * Usage: `./demos/quic/server/server <port> <certificate-file> <key-file>`

### doc/

Documentation organized in multiple formats:

* `doc/man1/` - Manual pages for command-line tools
* `doc/man3/` - Manual pages for API functions
* `doc/man5/` - Manual pages for file formats
* `doc/man7/` - Manual pages for overviews and concepts
* `doc/internal/` - Developer documentation

### include/

Header files that are installed along with OpenSSL:

* `include/openssl/` - Public API headers 
* `include/crypto/` - Headers for internal cryptographic implementations
* `include/internal/` - Headers for other internal components

### providers/

The provider architecture (new in OpenSSL 3.0) implementations:

* `providers/default/` - The default provider
* `providers/fips/` - The FIPS-validated module
* `providers/legacy/` - Legacy algorithms

### ssl/

Implementation of SSL, TLS, DTLS, and QUIC protocols:

* `ssl/record/` - SSL/TLS record layer
* `ssl/statem/` - SSL/TLS state machine
* `ssl/quic/` - QUIC protocol implementation (available since version 3.2.0)

### test/

Comprehensive test suite, organized by component and feature.

## Build System Files

> * `build.info` - Files in these directories define what gets built

Reviewer 2: You're explaining this file better in the patterns section further down, so this line is really not needed

```suggestion
```

* `Configurations/` - Platform-specific build configurations
* `exporters/` - Utilities for exporting OpenSSL build information to various build systems
* `os-dep/` - Operating system dependent code and platform-specific implementations
* `util/` - Build scripts and utilities

## External Modules

> OpenSSL can include several external modules as Git submodules, including:

Reviewer 2: ```suggestion
OpenSSL includes several external modules as Git submodules, mainly used for testing, including:
```

└── VMS/          # OpenVMS specific files
```

## Root-Level Files

The repository root contains many important files:

* `README.md` - The main README with project overview and links to documentation
* `INSTALL.md` - Detailed installation instructions for all platforms
* `Configure` - The main build configuration script
* `config` - A wrapper around Configure for Unix platforms
* `LICENSE.txt` - The Apache 2.0 license that applies to OpenSSL 3.0+
* `CHANGES.md` - A comprehensive changelog
* `NEWS.md` - A curated list of major changes in each release
* `VERSION.dat` - Contains the current version information
* `*.pc.in` and `configdata.pm.in` - Templates used during build configuration

Various platform-specific notes are also at the root level:
* `NOTES-*.md` - Platform-specific information (UNIX, WINDOWS, etc.)
* `README-*.md` - Additional READMEs for specific components 
  (ENGINES, FIPS, etc.)

## Major Directories

### apps/

Contains the OpenSSL command-line application source code. 
This is what you're using when you run the `openssl` command.

Notable components:
* The `openssl s_client` tool supports QUIC via the `-quic` flag 
  (since version 3.2.0)

> * The `openssl s_server` tool does NOT support QUIC directly

Reviewer 3: We mention QUIC here explicitly but not TLS which is slightly odd.

Author: @mattcaswell I've addressed your concern about the excessive detail in the demos section in commit fa95c4f76f 'Address review feedback on directory structure doc'.
I've simplified the QUIC server implementation details to just a single bullet point 'demos/quic/ - QUIC protocol examples' and added a clarification that these are 'some notable examples (not an exhaustive list)'.
The demos section is now much more concise and consistent with the level of detail throughout the rest of the document. Please let me know if you think it needs any further adjustments.

### crypto/

The heart of OpenSSL containing implementations of cryptographic algorithms. 

> Inside you'll find subdirectories for different cryptographic functions. 
> This is the code that ends up in `libcrypto`.

Reviewer 3: ```suggestion
Inside you'll find subdirectories for different cryptographic functions as well as other supporting code. 
This is the code that ends up in `libcrypto`. For example:
```

* `crypto/aes/` - AES encryption implementation
* `crypto/bn/` - Big Number arithmetic library
* `crypto/evp/` - High-level cryptographic API
* `crypto/rsa/`, `crypto/ec/`, etc. - Various algorithm implementations

### demos/

Example code demonstrating how to use OpenSSL APIs:

* `demos/quic/` - QUIC protocol examples, including a dedicated server 
  implementation
  * `demos/quic/server/` - Contains a standalone QUIC server application 
    that should be used instead of `s_server` for QUIC functionality

>   * Usage: `./demos/quic/server/server <port> <certificate-file> <key-file>`

Reviewer 3: This feels like a bit too much detail for this document

Author: Simplified the QUIC entry to just "demos/quic/ - QUIC protocol examples"
Removed the detailed server usage information

* `NEWS.md` - A curated list of major changes in each release
* `VERSION.dat` - Contains the current version information
* `*.pc.in` and `configdata.pm.in` - Templates used during build configuration

Various platform-specific notes are also at the root level:
* `NOTES-*.md` - Platform-specific information (UNIX, WINDOWS, etc.)
* `README-*.md` - Additional READMEs for specific components 
  (ENGINES, FIPS, etc.)

## Major Directories

### apps/

Contains the OpenSSL command-line application source code. 
This is what you're using when you run the `openssl` command.

Notable components:
* The `openssl s_client` tool supports QUIC via the `-quic` flag 
  (since version 3.2.0)
* The `openssl s_server` tool does NOT support QUIC directly

### crypto/

The heart of OpenSSL containing implementations of cryptographic algorithms. 
Inside you'll find subdirectories for different cryptographic functions. 
This is the code that ends up in `libcrypto`.

* `crypto/aes/` - AES encryption implementation
* `crypto/bn/` - Big Number arithmetic library
* `crypto/evp/` - High-level cryptographic API
* `crypto/rsa/`, `crypto/ec/`, etc. - Various algorithm implementations

### demos/

> Example code demonstrating how to use OpenSSL APIs:

Reviewer 3: If you're not going to give a full list of all sub-directories we should probably make it clear that this is just a sample of the things you can find here

> 

Reviewer 3: mention of `demos/guide` is probably also worthwhile, i.e. this is where all the code samples referred to from the ossl-guide-* man pages live

* `demos/quic/` - QUIC protocol examples, including a dedicated server 
  implementation
  * `demos/quic/server/` - Contains a standalone QUIC server application 
    that should be used instead of `s_server` for QUIC functionality
  * Usage: `./demos/quic/server/server <port> <certificate-file> <key-file>`

### doc/

Documentation organized in multiple formats:

* `doc/man1/` - Manual pages for command-line tools
* `doc/man3/` - Manual pages for API functions
* `doc/man5/` - Manual pages for file formats
* `doc/man7/` - Manual pages for overviews and concepts
* `doc/internal/` - Developer documentation

### include/

> Header files that are installed along with OpenSSL:

Reviewer 3: Well - only `include/openssl` gets "installed" (i.e. when you run `make install`). The other stuff is just internal and does not form part of the public API

* `include/openssl/` - Public API headers 
* `include/crypto/` - Headers for internal cryptographic implementations
* `include/internal/` - Headers for other internal components

### providers/

The provider architecture (new in OpenSSL 3.0) implementations:

> * `providers/default/` - The default provider

Reviewer 3: This directory does not exist!

Author: Updated with correct directories:
providers/common/ - Common provider code
providers/implementations/ - Algorithm implementations for various providers
providers/legacy/ - Legacy provider configuration

* `providers/fips/` - The FIPS-validated module

> * `providers/legacy/` - Legacy algorithms

Reviewer 3: Neither does this one!

Reviewer 3: > providers/legacy/ - Legacy provider configuration

There is no such directory!!

Author: @mattcaswell Resolved.

### ssl/

Implementation of SSL, TLS, DTLS, and QUIC protocols. 
This is the code that ends up in `libssl`.

* `ssl/record/` - SSL/TLS record layer
* `ssl/statem/` - SSL/TLS state machine
* `ssl/quic/` - QUIC protocol implementation (available since version 3.2.0)

### test/

> Comprehensive test suite, organized by component and feature.

Reviewer 3: Perhaps worthy of mention here is `test/recipes` where all the test recipes are held (it's these files that get invoked when you run `make test` - they then use the executables in the `test` directory. Also worthy of mention is `test/certs`

## Build System Files

> * `build.info` - Files in these directories define what gets built

Reviewer 3: build.info is not a directory - its a file

Author: Corrected to clarify these are files: "build.info files (found in many directories) describe what to build in that directory"

* `exporters/` - Utilities for exporting OpenSSL build information 
  to various build systems
* `os-dep/` - Operating system dependent code and platform-specific 
  implementations
* `util/` - Build scripts and utilities

## External Modules

OpenSSL includes several external modules as Git submodules, 
mainly used for testing, including:

* `oqs-provider/` - Post-quantum cryptography provider
* `krb5/` - Kerberos support
* `cloudflare-quiche/` - QUIC implementation from Cloudflare 
  (used as a reference for OpenSSL's own QUIC implementation)

## Development and Contributor Files

* `.ctags.d/` - Configuration for ctags, enabling better code navigation 
  and indexing for developers
* `.github/` - GitHub-specific files (workflows, issue templates)
* `CONTRIBUTING.md` - Guidelines for contributing to OpenSSL
* `HACKING.md` - Information on adding custom modifications to OpenSSL source
* `dev/` - Development utilities, helper scripts, and tools used during 
  OpenSSL development

## Stability Information

The API provided in the public headers (`include/openssl/`) is considered 
stable within major versions, while internal headers may change more 

> frequently. The provider interface is also considered stable.

Reviewer 3: Internal header files can change even between patch releases and does not form part of the public API

Author: Updated to clarify: "Internal headers (in include/crypto/ and include/internal/) may change even between patch releases and are not part of the public API"

OpenSSL Repository Structure
===========================

This document provides an overview of the OpenSSL repository's directory
structure, explaining what the major directories and files are, their purpose,
and how they relate to each other.

This documentation is available in multiple locations:

* As a man page when OpenSSL is installed on your system

> * As a web page on OpenSSL's official website (www.openssl.org)

Reviewer 3: This document will not be a man page or a web page (only the .pod files get automatically made into these)

* In the source distribution of OpenSSL in the `doc` directory

> * In a Git checkout at `doc/directory-structure.md`

Reviewer 3: These two are kind of the same thing really, i.e. part of the source distribution.

If you're reading this in a checkout of the OpenSSL source code, you're viewing

> the document in its original location.

Reviewer 3: Probably this is obvious and not needed to be said.
